### PR TITLE
Broken Sigmoid backward: Add test and mlop for Sigmoid

### PR DESCRIPTION
**Pull request by kunwar31**
Sigmoid backward is broken for most of negative numbers, and this also breaks 
tanh, gelu and quick_gelu 

for large negatives, in Exp backward, we get 
-0. in grad_output, inf in self.ret, MUL of which gives NaN as grad

Adding mlop for sigmoid solves the issue, but there is an oppurtunity for fusion here

### CODE REVIEW

**Comment by kunwar31** [comment]
torch sigmoid backward kernel for reference: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L725C6-L725C29

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/mlops.py                          9               0
tinygrad/tensor.py                         1               2
------------------------------------------------------------
total                                     10               2
------------------------------------------------------------
lines added in the tinygrad folder: 8
```

**Comment by geohot** [comment]
With a=-100 it doesn't test the normal case anymore

File: tinygrad/mlops.py [hunk 1]
```diff
@@ -55,6 +55,15 @@ def forward(self, x:LazyBuffer) -> LazyBuffer:
   def backward(self, grad_output:LazyBuffer) -> LazyBuffer:
     return self.ret.binary_op(BinaryOps.MUL, grad_output)
 
+class Sigmoid(Function):
+  __slots__ = "ret"
+  def forward(self, x:LazyBuffer) -> LazyBuffer:
+    self.ret = x.const_like(1).binary_op(BinaryOps.DIV, x.const_like(1).binary_op(BinaryOps.ADD, x.binary_op(BinaryOps.MUL, x.const_like(-math.log(math.e)/math.log(2))).unary_op(UnaryOps.EXP2)))
```
Comment (geohot) [inline tinygrad/mlops.py:61]: `math.log(math.e) = 1`

You can replace this in the other places too, it's just dumb.

**Comment by kunwar31** [comment]
> With a=-100 it doesn't test the normal case anymore

added the normal cases back

File: tinygrad/mlops.py [hunk 2]
```diff
@@ -138,7 +147,7 @@ def forward(self, x:LazyBuffer, y:LazyBuffer) -> LazyBuffer:
 
   def backward(self, grad_output:LazyBuffer):
     return grad_output.binary_op(BinaryOps.MUL, self.y.binary_op(BinaryOps.MUL, self.ret.binary_op(BinaryOps.DIV, self.x))) if self.needs_input_grad[0] else None, \
-           grad_output.binary_op(BinaryOps.MUL, self.x.unary_op(UnaryOps.LOG2).binary_op(BinaryOps.MUL, self.x.const_like(math.log(2)/math.log(math.e))).binary_op(BinaryOps.MUL, self.ret)) if self.needs_input_grad[1] else None
+           grad_output.binary_op(BinaryOps.MUL, self.x.unary_op(UnaryOps.LOG2).binary_op(BinaryOps.MUL, self.x.const_like(math.log(2)/1)).binary_op(BinaryOps.MUL, self.ret)) if self.needs_input_grad[1] else None
```
Comment (geohot) [inline tinygrad/mlops.py:150]: Umm

File: tinygrad/mlops.py [hunk 3]
```diff
@@ -41,20 +41,29 @@ class Log(Function):
   __slots__ = "x"
   def forward(self, x:LazyBuffer) -> LazyBuffer:
     self.x = x
-    return x.unary_op(UnaryOps.LOG2).binary_op(BinaryOps.MUL, x.const_like(math.log(2)/math.log(math.e)))
+    return x.unary_op(UnaryOps.LOG2).binary_op(BinaryOps.MUL, x.const_like(math.log(2)/1))
```
Comment (geohot) [inline tinygrad/mlops.py:44]: Please don't just find and replace and think about this for a minute.

Comment (kunwar31) [reply tinygrad/mlops.py:44]: my bad

**Comment by geohot** [comment]
Cool, should unblock #1073

**Pull request merged**